core/drivers/clk/clk.c

```
@@ -0,0 +1,279 @@
+{
+	struct clk *parent;
+
+	if (refcount_inc(&clk->enabled_count))
+		return TEE_SUCCESS;
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_enable(parent);
+
+	if (clk->ops->enable) {
+		if (clk->ops->enable(clk))
+			return TEE_ERROR_GENERIC;
+	}
+
+	refcount_set(&clk->enabled_count, 1);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_enable(struct clk *clk)
+{
+	TEE_Result ret;
+
+	cpu_spin_lock(&clk->lock);
+	ret = clk_enable_no_lock(clk);
+	cpu_spin_unlock(&clk->lock);
+
+	return ret;
+}
+
+static void clk_disable_no_lock(struct clk *clk)
+{
+	struct clk *parent;
+
+	if (!refcount_dec(&clk->enabled_count))
+		return;
+
+	if (clk->ops->disable)
+		clk->ops->disable(clk);
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_disable(parent);
+}
+
+void clk_disable(struct clk *clk)
+{
+	cpu_spin_lock(&clk->lock);
+	clk_disable_no_lock(clk);
+	cpu_spin_unlock(&clk->lock);
+}
+
+static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
+{
+	unsigned long parent_rate = 0;
+
+	if (clk->ops->set_rate) {
+		if (clk->parent)
+			parent_rate = clk->parent->rate;
+
+		if (clk->ops->set_rate(clk, rate, parent_rate))
+			return TEE_ERROR_GENERIC;
+	}
+
+	clk_compute_rate(clk);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	cpu_spin_lock(&clk->lock);
+	if (rate == clk->rate)
+		goto out;
+
+	if (clk->flags & CLK_SET_RATE_GATE) {
+		if (clk_is_enabled(clk)) {
+			ret = TEE_ERROR_BAD_STATE;
+			goto out;
+		}
+	}
+
+	ret = clk_set_rate_no_lock(clk, rate);
+out:
+	cpu_spin_unlock(&clk->lock);
+
+	return ret;
+}
+
+static TEE_Result clk_set_parent_no_lock(struct clk *clk, unsigned int pidx)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	bool was_enabled = false;
+
+	/* Requested parent is already the one set */
+	if (clk->parent == clk->parents[pidx])
+		return TEE_SUCCESS;
+
+	was_enabled = clk_is_enabled(clk);
+	if (was_enabled)
+		clk_disable_no_lock(clk);
```

this does not ensure the clock is disabled, there maybe other refcounts accumulated. This should return an error, as done in `clk_set_rate()`

---

This call to disable is meant to maintain the refcount to a correct value not to disable the clock. If the clock needs to be disabled while being modified, the driver should handle that. But CLK_SET_RATE_GATE is there for that kind of scenarios where the clock should not be modified with enable peripheral.

---

ok, i understand. thanks.
I suggest an short inline comment like `/* Decrement refcounts while on old parent clock */`
and `/* Re-increment refcounts with the new parent clock */` at line 253.

Or maybe no inline comment but rename variable `was_enabled` to `update_refcounts`.

---

I guess the comment is more appropriate here, I'll update that.